src/bidiMapper/modules/context/NavigationTracker.ts

```diff
@@ -0,0 +1,214 @@
+/*
+ *  Copyright 2024 Google LLC.
+ *  Copyright (c) Microsoft Corporation.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+import {
+  type BrowsingContext,
+  ChromiumBidi,
+} from '../../../protocol/protocol.js';
+import {Deferred} from '../../../utils/Deferred.js';
+import {getTimestamp} from '../../../utils/Time.js';
+import {urlMatchesAboutBlank} from '../../../utils/UrlHelpers.js';
+import {uuidv4} from '../../../utils/uuid.js';
+import type {EventManager} from '../session/EventManager.js';
+
+export type NavigationEventName =
+  // TODO: implement.
+  // ChromiumBidi.BrowsingContext.EventNames.DownloadWillBegin|
+  | ChromiumBidi.BrowsingContext.EventNames.FragmentNavigated
+  | ChromiumBidi.BrowsingContext.EventNames.NavigationAborted
+  | ChromiumBidi.BrowsingContext.EventNames.NavigationFailed;
+
+class NavigationState {
+  started = new Deferred<void>();
+  finished = new Deferred<NavigationEventName>();
+
+  // Tracks CDP events on the navigation. Used to distinguish new navigations from the
+  // initial one.
+  cdpStates = {
+    frameScheduledNavigation: false,
+    frameRequestedNavigation: false,
+    frameStartedLoading: false,
+  };
+
+  markFragmentNavigated() {
+    this.started.resolve();
```

Comment:
so fragment navigations should not have other navigation events like navigationStarted as per spec.